metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -135,18 +149,46 @@ class CheckpointingService[F[_]: Sync: Applicative, N](
     storeRunner.runReadWrite {
       ledgerStorage.put(ledger)
     } >>
-      ledgerTree.set(Map(header.hash -> (ledger, header))) >>
-      lastExecutedHeader.set(header)
+      ledgerTree.set(Map(header.hash -> (ledger, header, None))) >>
+      lastCommittedHeader.set(header)
   }
 
-  private def constructCheckpoint(
-      ledger: Ledger,
+  private def constructCheckpointCert(
       commitQC: QuorumCertificate[CheckpointingAgreement]
-  ): F[Option[CheckpointCertificate]] =
-    ??? //TODO: PM-3110
+  ): F[Option[CheckpointCertificate]] = {
+
+    def traverseCommitPath(
+        ledgers: LedgerTree,
+        from: Block.Hash,
+        path: List[Block.Header]
+    ): Option[CheckpointCertificate] =
+      ledgers.get(from) match {
+        case Some((_, h, Some(cp))) =>
+          val headers = NonEmptyList.of(h, path: _*)
+          CheckpointCertificate(headers, cp.checkpoint, cp.proof, commitQC).some
```

Comment:
Regarding:

> As we discussed it would be good to have tests for the creation of a checkpoint certificate, and it could live outside the service, for example in the CheckpointCertificate companion, given all the input values to construct the proof.

(possibly a longer discussion, so moving it here)

It's easy to move the proof construction to the companion object. But we still need to traverse the headers to construct the certificate. We could make this construction dependent on a function that returns an optional parent header, but we'd also have to verify the merkle proof on each step as final traversal condition.

Do you think we should do that?

Reply:
I wouldn't start passing functions for optional parent headers, I was more thinking about a pure function.

TBH the way you do it is so different from what I had in mind originally that it's not worth moving around. What I was thinking of was to have a buffer where you accumulate the inputs from the last block you encountered that had a checkpoint candidate to the block which has a commit, and then you only produce a proof at the end, so it would be something like `case class CheckpointCertificateInputs(lastBlockWithCheckpointCandidate: Block, blocksSinceLastCheckpointCandidate: List[Block.Header)`. 

Then it would be easy to just pass this block and the headers, along with the commit certificate, to a prover, it wouldn't have to retrieve anything else.

But because you are constructing a checkpoint proof for every block you see, and later go back to find the last one that had a non-empty proof in it, the logic is meshed into the whole service, not as clean to remove. So I wouldn't bother with moving just an iteration logic that has to retrieve headers until it finds something interesting.

Reply:
Well, it would be difficult to maintain an accumulated list. Unless we want to differentiate `updateLedger` between validation and execution we need to maintain a tree, as we already do. In either case the logic of building the path part of the proof would reside mostly in `CheckpointingService`. So really the logic of the potential smart constructor boils down to constructing the merkle proof.

I agree it would be nice to have something like `CheckpointCertificate.construct(block, headers, commitQC)`. The merkle proof would be constructed in that method. However, that would require to keep potentially multiple full blocks in memory (different branches) or looking up the block in storage. I wanted to avoid either.

I'd still put the merkle proof construction in a common place like `CheckpointCertificate` so that I can reuse it in tests.

Reply:
Exactly, I thought this list accumulation would only happen in execution, not validation, it only makes sense there. Only the headers would have been kept in memory and the body of the last block that had a checkpoint. 

But I don't want to upset your scheme. Moving the merkle proof construction to a reusable place is a good step.

Reply:
Ok, you're right. Duh, I can move the accumulation outside of `updateLedger`. I will do that

Reply:
Done, the checkpoint-relevant accumulation only happens in `executeBlock` now, and certificate construction code has been moved to the companion